Add configuration options for hosted plugin runtime

justinas opened the pull request:
Adds plugins flag to `Features`, adds configuration options to run the plugin runtime to the auth server config.

This configuration will be utilized by code in `e`.

Commit:
Add Plugins feature flag

Commit:
Add config for plugins runtime

justinas commented on lib/config/fileconf.go:
```diff
@@ -947,6 +948,9 @@ type Auth struct {
 	LoadAllCAs bool `yaml:"load_all_cas,omitempty"`
+
+	// Plugins configures the hosted plugins runtime
+	Plugins Plugins `yaml:"plugins,omitempty"`
```
Not sure if this is best place to put it this, but my reasoning is: `plugins` is not really like the other standalone services - it does not have a `listen_addr`, and can only run inside Auth.

r0mant replied:
If we want to be able to reuse the same configuration format for both standalone and in-auth modes, I'm not sure placing it under "auth" makes a lot of sense tbh. Why not a separate top-level section? Then we could reuse it for the new OpsGenie plugin too and have it be configured using same YAML fileconfig format rather than TOML.

And if we ever decide we want to make plugins into proper Teleport services (which I still think makes sense), it would be easier as well.

justinas committed:
Merge branch 'master' into justinas/plugin-config

ravicious commented:
I'll take a look at this PR tomorrow. But I'd recommend asking for another group 1 review anyway since I don't feel super confident when it comes to reviewing backend changes which affect user facing APIs, in this case the "API" being the config file.

ravicious reviewed (commented):
Looks good to me, but as I mentioned, I don't know if there are any extra precautions we need to take when adding new fields to the config file.

justinas commented:
@EdwardDowling and others: Let's discuss whether this can cover the standalone (non-hosted/non-cloud) plugins use-case as well. Ref:

#22221 
https://github.com/gravitational/teleport/pull/22221#discussion_r1119098424

Commit:
Allow reading OAuth secrets from files

Commit:
Merge remote-tracking branch 'origin/master' into justinas/plugin-config

Commit:
Rename Plugins to HostedPlugins

Make it clear that this is separate from the upcoming
self-hosted, standalone plugins that will be part of the Teleport
binary.

justinas commented:
Hey folks,

I've updated the code to be able to deal with the secrets in a more secure way (allow reading from a file). I've also renamed the field to `hosted_plugins` to make it a bit more clear that this is distinct from the planned standalone `plugins`.

I am still unsure about merging the two structures, as they serve different purposes: in Cloud, this only sets up the static "plumbing", i.e. auth secrets and similar parameters that *we* need to implement the plugins. In standalone plugins, all of the interesting, plugin-specific settings are also defined in the `plugins` YAML config.

Considering this (hosted plugins) will be a Cloud-specific feature and end users of self-hosted Teleport should not interact with this for the time being, we will have some degree of freedom to revisit this later on without much breakage. If my reasoning makes sense, let's merge this sometime soon 🙂

r0mant commented on lib/config/fileconf.go:
```diff
@@ -982,6 +983,10 @@ type Auth struct {
+
+	// HostedPlugins configures the hosted plugins runtime.
+	// This is currently Cloud-specific.
+	HostedPlugins HostedPlugins `yaml:"hosted_plugins,omitempty"`
```
Since we're going to have a separate `plugin_service` (see OpsGenie RFD), can we make this a part of its configuration right away?

```yaml
plugin_service:
  enabled: yes
  opsgenie:
    api_key: /home/ubuntu/key
  slack:
    client_id:
    client_secret:
```

r0mant replied:
As discussed in Slack, self-hosted versions of Slack plugin can keep using auth tokens so this configuration is specific to Cloud's hosted plugins use-case and can probably stay here for now.

r0mant commented on lib/config/fileconf.go:
```diff
@@ -1300,6 +1305,67 @@ func (dt *DeviceTrust) Parse() (*types.DeviceTrust, error) {
+		}
+		clientID = strings.TrimSpace(string(content))
+	}
+	if strings.HasPrefix(clientSecret, "/") {
```
Should we disallows anything that's not a path? I think our approach has been to now allow embedding secrets in the configs and only support reading from files.

r0mant commented on lib/config/fileconf.go:
```diff
@@ -1300,6 +1305,67 @@ func (dt *DeviceTrust) Parse() (*types.DeviceTrust, error) {
+
+func (o *OAuthClientCredentials) Parse() (*oauth2.ClientCredentials, error) {
+	if o.ClientID == "" || o.ClientSecret == "" {
+		return nil, trace.BadParameter("Both client_id and client_secret must be specified")
```
```suggestion
		return nil, trace.BadParameter("both client_id and client_secret must be specified")
```

r0mant reviewed (approved):
lgtm once you update config parsing to always treat client_id and client_secret as paths.

Commit:
Only allow to read client credentials from files

justinas committed:
Merge branch 'master' into justinas/plugin-config
